program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
+        let mut entry = current_entry;
+        for _ in 0..num_entries {
+            self.entry_validity[entry] = 0;
+            entry = Self::get_next_entry(entry);
+        }
+    }
+    /// returns a weighted average of v with weights w
+    fn weighted_average<
+        T: std::ops::Mul<Output = T>
+            + std::ops::Add<Output = T>
+            + std::ops::Div<Output = T>
+            + std::marker::Copy,
+    >(
+        w1: T,
+        v1: T,
+        w2: T,
+        v2: T,
+    ) -> Result<T, OracleError> {
+        Ok((w1 * v1 + w2 * v2) / (w1 + w2))
+    }
+    ///Adds price and confidence each multiplied by update_time
+    /// to the entry
+    fn add_price_to_entry(
+        &mut self,
+        prev_price: i64,
+        prev_conf: u64,
+        current_price: i64,
+        current_conf: u64,
+        update_time: i64,
+        entry: usize,
+    ) -> Result<(), OracleError> {
+        let avg_price = (prev_price + current_price) / 2;
+        let avg_conf = (prev_conf + current_conf) / 2;
+        self.running_price[entry] +=
+            try_convert::<i64, SignedTrackerRunningSum>(update_time * avg_price)?;
+        self.running_confidence[entry] += try_convert::<u64, UnsignedTrackerRunningSum>(
+            try_convert::<i64, u64>(update_time)? * avg_conf,
+        )?;
+        //since price is always greater than confidence, it is sufficient to look at the
+        // price we also look at the average because it is the value that
+        // gets multiplied with the number of seconds
+        self.max_observed_entry_abs_val =
+            cmp::max(try_convert(avg_price)?, self.max_observed_entry_abs_val);
+        Ok(())
+    }
+}
+
+impl<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64>
+    Tracker<GRANUALITY, NUM_ENTRIES, THRESHOLD> for SmaTracker<GRANUALITY, NUM_ENTRIES, THRESHOLD>
+{
+    fn initialize(&mut self) -> Result<(), OracleError> {
+        //ensure that the first entry is invalid
+        self.max_update_time_gap = THRESHOLD;
+        Ok(())
+    }
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError> {
+        //multiply by precision multiplier so that our entry aggregated can be rounded to have
+        //the same accuracy as user input
+        let inflated_prev_price = prev_price * SMA_TRACKER_PRECISION_MULTIPLIER;
+        let inflated_current_price = current_price * SMA_TRACKER_PRECISION_MULTIPLIER;
+        let inflated_prev_conf =
+            prev_conf * try_convert::<i64, u64>(SMA_TRACKER_PRECISION_MULTIPLIER)?;
+        let inflated_current_conf =
+            current_conf * try_convert::<i64, u64>(SMA_TRACKER_PRECISION_MULTIPLIER)?;
+
+        let update_time = current_time - prev_time;
```

Comment:
can you name this `update_time_delta` or something that clearly indicates that this is an interval of time, not a point?